### PR TITLE
feature/SAUTHC-2_Add_refresh_token_feature

### PR DESCRIPTION
**Pull request by Abdallah-Abdelkhalek**
Still unsure about the Request & Response parameters so please check.

### CODE REVIEW

File: src/AuthManager.ts [hunk 1]
```diff
@@ -50,6 +50,40 @@ export class AuthManager {
...
+                    headers: {
+                        'Content-Type': 'application/json',
+                    },
+                    body: JSON.stringify({
+                        refresh_token: refreshToken,
+                    }),
+                })
+                    .then((response) => {
+                        if (response.status !== 200) {
+                            throw new Error('Failed to refresh the token');
```
Comment (amrography) [inline src/AuthManager.ts:68]: You can use the `reject()` here. and remove the `try-catch` block

**Review by amrography — APPROVED** [review]

File: src/AuthManager.ts [hunk 2]
```diff
@@ -50,6 +50,40 @@ export class AuthManager {
         return { newCodeVerifier, newCodeChallenge };
     };
 
+    private async refreshAccessToken(): Promise<void> {
+        return new Promise((resolve,reject) => {
+            try {
+                const refreshToken = localStorage.getItem('refresh_token');
+                fetch(`${this.authServer}auth/refresh`, {
```
Comment (amrography) [inline src/AuthManager.ts:57]: If `axios` wasn't used then we should remove it

File: src/AuthManager.ts (hunk 1 above)
Comment (HassanElZarkawy) [reply src/AuthManager.ts:68]: Try/catch is important, as it also protect against network errors

**Review by HassanElZarkawy — APPROVED** [review]

**Review by MostafaFawzy7 — APPROVED** [review]

File: src/AuthManager.ts [hunk 3]
```diff
@@ -88,9 +122,8 @@ export class AuthManager {
                 if (decodedToken) {
                     const currentTime = Date.now() / 1000;
                     if (decodedToken.exp < currentTime) {
-                        // add refresh check here instead and
-                        localStorage.removeItem('access_token');
-                        return resolve(false);
+                        // add refresh check here
+                        this.refreshAccessToken();
```
Comment (SirNarsh) [inline src/AuthManager.ts:126]: refresh token also has expiry time, check expiry of refresh token before trying to use it

File: src/AuthManager.ts [hunk 4]
```diff
@@ -115,9 +148,8 @@ export class AuthManager {
                 if (decodedToken) {
                     const currentTime = Date.now() / 1000;
                     if (decodedToken.exp < currentTime) {
-                        // add refresh check here instead and
-                        localStorage.removeItem('access_token');
-                        throw new Error('Access token expired');
+                        // add refresh check here
+                        this.refreshAccessToken();
```
Comment (SirNarsh) [inline src/AuthManager.ts:152]: can you apply DRY on `getAccessToken` `isLoggedIn`

File: src/AuthManager.ts [hunk 5]
```diff
@@ -50,6 +49,74 @@ export class AuthManager {
...
+                    },
+                    body: JSON.stringify({
+                        refresh_token: refreshToken,
+                    }),
+                })
+                    .then((response) => {
+                        if (response.status !== 200) {
+                            throw new Error('Failed to refresh the token');
+                        }
+                        return response.json();
```
Comment (HassanElZarkawy) [inline src/AuthManager.ts:79]: You can't return inside a promise, you need to `resolve(response.json())`

File: src/AuthManager.ts [hunk 6]
```diff
@@ -50,6 +49,74 @@ export class AuthManager {
...
+                // decode access token and check if it's expired
+                const decodedToken = accessToken
+                    ? JSON.parse(atob(accessToken.split('.')[1]))
+                    : null;
+                if (decodedToken) {
+                    const currentTime = Date.now() / 1000;
+                    if (decodedToken.exp < currentTime) {
+                        // refreshing expired token
+                        const newAccessToken = await this.refreshAccessToken();
+                        return newAccessToken;
```
Comment (HassanElZarkawy) [inline src/AuthManager.ts:110]: You can't return inside a promise, you need to `resolve(newAccessToken)`

**Review by amrography — APPROVED** [review]

File: src/AuthManager.ts [hunk 7]
```diff
@@ -75,25 +142,12 @@ export class AuthManager {
...
-                if (decodedToken) {
-                    const currentTime = Date.now() / 1000;
-                    if (decodedToken.exp < currentTime) {
-                        // add refresh check here instead and
-                        localStorage.removeItem('access_token');
-                        return resolve(false);
-                    }
-                }
-
+                this.checkAccessToken();
```
Comment (amrography) [inline src/AuthManager.ts:150]: should you use `await` ?

Comment (Abdallah-Abdelkhalek) [reply src/AuthManager.ts:150]: yess correct, Done 👍

File: src/AuthManager.ts [hunk 8]
```diff
@@ -50,6 +49,74 @@ export class AuthManager {
...
+                        resolve(exchangeJson.access_token);
+                    })
+                    .catch((error) => {
+                        reject(error);
+                    });
+            } catch (error) {
+                reject(error);
+            }
+        });
+    }
```
Comment (SirNarsh) [inline src/AuthManager.ts:93]: missing empty line between methods, do we have eslint-supaapps installed here?

Comment (Abdallah-Abdelkhalek) [reply src/AuthManager.ts:93]: no linting & no prettier Should i add them in this PR?

**Pull request merged**